controllers/virtualmachinereplicaset/virtualmachinereplicaset_controller.go

```diff
@@ -202,38 +220,462 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Re
+	// since syncing might have resulted in replicas being added or removed.
+	r.updateStatus(ctx, ctx.ReplicaSet, filteredVMs)
+
+	if syncErr != nil {
+		return ctrl.Result{}, errors.Wrapf(syncErr, "failed to sync VirtualMachineReplicaSet replicas")
+	}
+
+	var replicas int32
+	if ctx.ReplicaSet.Spec.Replicas != nil {
+		replicas = *ctx.ReplicaSet.Spec.Replicas
+	}
+
+	// Queue faster reconciles until all replicas are available.
+	if ctx.ReplicaSet.Status.ReadyReplicas != replicas {
+		return ctrl.Result{RequeueAfter: 15 * time.Second}, nil
+	}
+
+	return ctrl.Result{}, nil
 }
 
-func (r *Reconciler) ReconcileDelete(ctx *pkgctx.VirtualMachineReplicaSetContext) (reterr error) {
-	if controllerutil.ContainsFinalizer(ctx.ReplicaSet, finalizerName) {
-		defer func() {
-			r.Recorder.EmitEvent(ctx.ReplicaSet, "Delete", reterr, false)
-		}()
+// MustEqualValue returns true if the replica set name equals either the label
+// value or its hashed value.
+func MustEqualValue(rsName, labelValue string) bool {
+	return labelValue == util.MustFormatValue(rsName)
+}
 
-		// Handle deletion of replicas
+// getNewVirtualMachine creates a new VirtualMachine object. The name of the newly created resource is going
+// to be created by the API server, we set the generateName field.
+func (r *Reconciler) getNewVirtualMachine(rs *vmopv1.VirtualMachineReplicaSet) *vmopv1.VirtualMachine {
+	// gv := vmopv1.SchemeGroupVersion
+	vm := &vmopv1.VirtualMachine{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: fmt.Sprintf("%s-", rs.Name),
+			// Note: by setting the ownerRef on creation we signal to the
+			// VirtualMachine controller that this is not a stand-alone VM.
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(rs, replicaSetKind)},
+			Namespace:       rs.Namespace,
+			Labels: map[string]string{
+				vmopv1.VirtualMachineReplicaSetNameLabel: util.MustFormatValue(rs.Name),
+			},
+			Annotations: rs.Spec.Template.Annotations,
+		},
+		// TypeMeta: metav1.TypeMeta{
+		// 	Kind:       gv.WithKind("VirtualMachine").Kind,
+		// 	APIVersion: gv.String(),
+		// },
+		Spec: rs.Spec.Template.Spec,
+	}
 
-		controllerutil.RemoveFinalizer(ctx.ReplicaSet, finalizerName)
+	vm.Labels = getLabelsFromVMReplicaSet(rs)
+
+	vm.Annotations = getAnnotationsFromReplicaSet(rs)
+
+	// TODO: Propagate the VirtualMachineDeploymentNameLabel from VirtualMachineReplicaSet to VirtualMachines if it exists.
+
+	return vm
+}
+
+func getLabelsFromVMReplicaSet(rs *vmopv1.VirtualMachineReplicaSet) map[string]string {
+	labels := make(map[string]string)
+
+	// Copy over the labels from the template
+	maps.Copy(labels, rs.Spec.Template.Labels)
+
+	// Ensure that the ReplicaSetLabel is always present.
+	labels[vmopv1.VirtualMachineReplicaSetNameLabel] = util.MustFormatValue(rs.Name)
+
+	return labels
+}
+
+func getAnnotationsFromReplicaSet(rs *vmopv1.VirtualMachineReplicaSet) map[string]string {
+	annotations := make(map[string]string)
```

Comment:
This could be optimized as such `annotations := make(map[string]string, len(rs.Spec.Template.Annotations)` so that it is already sized according to the needs of the copy below.